### PR TITLE
Add Enterprise Service

### PR DESCRIPTION
**Pull request by timraymond**
This restores the registration functionality that was removed in 718fb92c and retrofits it to work for Enterprise. When configured, nodes register with one or more instances of Enterprise, and begin periodically reporting stats to them. An http service is also started on a configureable port which will  facilitate the enterprise manager's administrative capabilities.

### CODE REVIEW

File: services/enterprise/service.go [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+		RegisterDiagnosticsClient(name string, client monitor.DiagsClient)
+	}
+
+	enabled       bool
+	hosts         []*client.Host
+	statsInterval time.Duration
+	version       string
+	mu            sync.Mutex
+	lastContact   time.Time
+	adminPort     string
```
Comment (e-dard) [inline services/enterprise/service.go]: Can you add a space above `adminPort` and below `version`, to indicate that `mu` is protecting `lastContact`?

**Comment by timraymond** [comment]
@e-dard Updated! Also, I've added some commits to address a race condition for the token and secret when setting up the admin API.

**Comment by corylanou** [comment]
Need a rebase.

**Comment by timraymond** [comment]
Rebased

**Comment by timraymond** [comment]
@jwilder I made the changes we discussed last week in 0201ca6

File: services/enterprise/service.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+	clusterID := s.MetaStore.ClusterID()
+
+	t := time.NewTicker(s.statsInterval)
+	for {
+		select {
+		case <-t.C:
+			stats, err := s.Monitor.Statistics(nil)
+			if err != nil {
+				s.logger.Printf("failed to retrieve statistics: %s", err.Error())
```
Comment (e-dard) [inline services/enterprise/service.go]: nit: should be `Failed`

File: services/enterprise/service.go [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+					Name:   x.Name,
+					Tags:   x.Tags,
+					Values: x.Values,
+				}
+			}
+			st.Data = data
+
+			resp, err := cl.Save(st)
+			if err != nil {
+				s.logger.Printf("failed to post statistics to Enterprise: repsonse code: %d: error: %s", resp.StatusCode, err)
```
Comment (e-dard) [inline services/enterprise/service.go]: as above

**Comment by e-dard** [comment]
couple of nits but LGTM :+1

**Comment by timraymond** [comment]
Nits fixed—lmk if I should squash.

**Comment by e-dard** [comment]
@timraymond I guess you could squash this down and then feel free to merge.

File: services/enterprise/service.go [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+// SetLogger sets the internal logger to the logger passed in.
+func (s *Service) SetLogger(l *log.Logger) {
+	s.logger = l
+}
+
+// Diagnostics returns diagnostics information.
+func (s *Service) Diagnostics() (*monitor.Diagnostic, error) {
+	diagnostics := map[string]interface{}{
+		"Hosts":        s.hosts,
+		"last_contact": s.getLastContact().String(),
```
Comment (jwilder) [inline services/enterprise/service.go]: Should probably be `lastContact` to match naming convention of other stats.

File: services/enterprise/service.go [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+// SetLogger sets the internal logger to the logger passed in.
+func (s *Service) SetLogger(l *log.Logger) {
+	s.logger = l
+}
+
+// Diagnostics returns diagnostics information.
+func (s *Service) Diagnostics() (*monitor.Diagnostic, error) {
+	diagnostics := map[string]interface{}{
+		"Hosts":        s.hosts,
```
Comment (jwilder) [inline services/enterprise/service.go]: Should be `hosts` I think.

**Comment by jwilder** [comment]
Two smalls things but, :+1:  after that.

**Comment by jwilder** [comment]
```
services/enterprise/service.go:94: cannot call non-function s.lastContact (type time.Time)
services/enterprise/service.go:203: type Service has both field and method named lastContact
```

**Pull request closed without merging**